source/secure_tunneling.c

```diff
@@ -70,15 +73,19 @@ static void s_on_websocket_setup(
     secure_tunnel->websocket = websocket;
     secure_tunnel->config.on_connection_complete(secure_tunnel->config.user_data);
 
-    struct aws_event_loop *event_loop =
-        aws_event_loop_group_get_next_loop(secure_tunnel->config.bootstrap->event_loop_group);
-    aws_event_loop_schedule_task_now(event_loop, &secure_tunnel->ping_task);
+    secure_tunnel->ping_task_context->task_cancelled = false;
+    secure_tunnel->ping_task_context->websocket = websocket;
+    aws_event_loop_schedule_task_now(secure_tunnel->ping_task_context->event_loop, &secure_tunnel->ping_task);
 }
 
 static void s_on_websocket_shutdown(struct aws_websocket *websocket, int error_code, void *user_data) {
     UNUSED(websocket);
     UNUSED(error_code);
-    UNUSED(user_data);
+
+    struct aws_secure_tunnel *secure_tunnel = user_data;
+    secure_tunnel->ping_task_context->task_cancelled = true;
```

Comment:
I want to confirm: do you mean:

1. Cancel the task here? Or 
2. Cancel the task within `s_ping_task`? 

For Option 1, @DavidOgunsAWS and I talked about this. Please see this [comment](https://github.com/awslabs/aws-c-io/blob/main/include/aws/io/event_loop.h#L289). I wasn't sure what does it mean. But David confirmed that the `aws_event_loop_cancel_task` must be called from the same thread that the task was scheduled. It cannot be certain that the thread running this function is the same thread as when I schedule the task in `s_on_websocket_setup`

Reply:
I want to confirm: do you mean:

1. Cancel the task here? Or 
2. Cancel the task within `s_ping_task`? 

For Option 1, @DavidOgunsAWS and I talked about this. Please see this [comment](https://github.com/awslabs/aws-c-io/blob/main/include/aws/io/event_loop.h#L289). I wasn't sure what does it mean. But David confirmed that the `aws_event_loop_cancel_task` must be called from the same thread that the task was scheduled. It cannot be certain that the thread running this function is the same thread as when I schedule the task in `s_on_websocket_setup`

Another question: by not scheduling the task again (see `s_ping_task`), do I still have to cancel it?

Reply:
I have made `task_cancelled` an `aws_atomic_var`. I also call `aws_event_loop_cancel_task` within `s_ping_task`.

Reply:
Right now it's very hard to offer advice here because there's not a good way to reason about the secure tunnel type's lifecycle (relative to the things it depends on, as well as what might depend on it) given the current approach.

Around Aug/Sep we moved the persistent async types to a ref-counted model (your stuff may have already been in progress by that point and it could have been easily missed) and over time we've transitioned types that have async shutdown behavior to have configurable shutdown callbacks so that users or depending types can be notified properly when something is truly gone (in turn allowing them to clean up at the proper time).  aws_global_thread_creator_shutdown_wait() is useful shorthand for tests and at the end of programs, but shouldn't be a crutch for the shutdown of individual types (even if they're intended as singletons).

For the ping task, we have a very similar situation in the http connection manager (https://github.com/awslabs/aws-c-http/blob/main/source/connection_manager.c#L248), where the manager has a periodic task that culls idle connections.  Correct tear down of the manager and its task is actually pretty tricky and complicated.  In that case we have a multi-phase shutdown and we schedule a task on the target event loop to do the cancellation of the periodic task.  Completion of that cancellation task is one of the required phases of shutdown.

Here's my rough recommendation list to move aws_secure_tunnel in line with other async types (obviously this goes way beyond the scope of your code review, but I think it's important and I'm certainly willing to help work through these):

1. Make it ref-counted (add an acquire function, modify release appropriately), use the helper type in https://github.com/awslabs/aws-c-common/blob/main/include/aws/common/ref_count.h and see client_bootstrap, event_loop_group, etc... for usage.
1. In new(), take and hold refs to client_bootstrap, tls_context, copy config cursors into held strings
1. Move to an async shutdown process that release (with no remaining refs) triggers the start of
1. Release the refs on client_bootstrap and tls_context during the shutdown process once you know your usage of them has completed (potentially when the websocket has completed shutdown or when the ping task is cancelled whichever is later)
1. Remove the task_cancelled bool and use a one-off scheduled task to cancel the ping task, and once cancelled, trigger the next phase of shutdown
1. Add a (optional) config shutdown callback that gets invoked after everything is cleaned up

So shutdown looks like it would follow this sequence:

1. Ref count hits zero
1. Start shutdown of websocket (or skip to next step if websocket is already shutdown)
1. On shutdown completion for websocket, submit cancel task for ping
1. On cancel task for ping (either cancelled or run), trigger final shutdown
1. Cleanup local state, release references to client_bootstrap and tls_context, release secure tunnel struct memory
1. Invoke the optional shutdown callback if it was set

Finally, if you want (probably) to hold off on a big ref-counting/shutdown revamp and just get your cancelled ping task in asap, then I think you could potentially just add the cancellation task and then once it completes invoke the memory cleanup that you currently do in release (switching release to just scheduling the cancel task).